components/push/src/storage/db.rs

```diff
@@ -62,6 +62,28 @@ impl PushDb {
         let conn = Connection::open_in_memory()?;
         Ok(Self::with_connection(conn)?)
     }
+
+    /// Normalize UUID values to undashed, lowercase.
+    // The server mangles ChannelID UUIDs to undashed lowercase values. We should force those
+    // so that key lookups continue to work.
+    pub fn normalize_uuid(uuid: &str) -> String {
+        uuid.replace('-', "").to_lowercase()
+    }
+
+    /// Dash UUID strings.
+    // In case it's needed.
+    pub fn uuid_to_dashed(uuid: &str) -> Result<String> {
+        let unorm = Self::normalize_uuid(uuid);
+        let norm = unorm.as_bytes();
+        Ok(format!(
+            "{}-{}-{}-{}-{}",
+            String::from_utf8(norm[0..8].to_vec()).unwrap_or_else(|_| "xxxxxxxx".to_owned()),
```

Comment:
This is rather strange. Shouldn't it just be `format!("{}-{}-{}-{}-{}", unorm[0..8], unorm[8..12], ...etc)`?

As it is we're getting the underlying buffer as bytes, then converting it back to a string manually.

I guess a difference is that this version will handle the case where both:
- the uuid has non-ascii characters
- one of the multibyte utf8 sequences falls across the `8, 12, 16, 20` boundaries.

But probably that doesn't matter (honestly I'd support a `assert!(uuid.is_ascii())` here or in normalize_uuid).

Reply:
1) uuids should never have non-ascii characters. If they do, and things break, that's a good thing since the UUID would be invalid. (UUIDs are just oddly formatted hex representations of a 128 bit number)

2) honestly, it's probably far better to use something like [uuid](https://www.crates.io/crates/uuid) for serious UUID manipulation. The normalization (remove dashes, drop case) is used, the "de-normalize" is more a hack for testing and printing.

I'll add a check to enforce `uuid.is_ascii()` and return an error.

Reply:
Err, to be more concrete, is there a reason that this wouldn't work:
```rust
    pub fn uuid_to_dashed(uuid: &str) -> Result<String> {
        let norm = Self::normalize_uuid(uuid);
        Ok(format!(
            "{}-{}-{}-{}-{}",
            &norm[0..8],
            &norm[8..12],
            &norm[12..16],
            &norm[16..20],
            &norm[20..]
        ))
    }
```

Reply:
note: this would be in addition to a check for is_ascii / possibly the len, etc.

Reply:
No. Heh, I had forgotten to borrow then normalized uuid and apparently decided to do the worst possible replacement version. (d'oh)